File: ErmisClient/Mobile/ermis_client/lib/core/networking/handlers/client_message_handler.dart
/* Copyright (C) 2025 Ilias Koukovinis <ilias.koukovinis@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

import 'dart:convert';
import 'dart:typed_data';
import 'package:ermis_mobile/core/networking/common/message_types/content_type.dart';
import 'package:ermis_mobile/core/networking/common/message_types/message_delivery_status.dart';
import 'package:ermis_mobile/core/data/models/network/byte_buf.dart';
import 'package:ermis_mobile/core/models/message_events.dart';
import 'package:ermis_mobile/core/models/chat_session.dart';
import 'package:ermis_mobile/core/models/message.dart';
import 'package:ermis_mobile/core/networking/user_info_manager.dart';
import '../../event_bus/app_event_bus.dart';

final AppEventBus _eventBus = AppEventBus.instance;

class ClientMessageHandler {
  static void handle(ByteBuf msg) {
    MessageContentType? contentType = MessageContentType.fromId(msg.readInt32());
    int epochSecond = msg.readInt64();

    Map<MessageFields, Uint8List?> fields = {};
    switch (contentType) {
      case MessageContentType.text || MessageContentType.gif:
        var textLength = msg.readInt32();
        Uint8List? text = msg.readBytes(textLength);
        fields[MessageFields.text] = text;
        break;
      case MessageContentType.file ||
            MessageContentType.image ||
            MessageContentType.voice ||
            MessageContentType.video:
        var fileNameLength = msg.readInt32();
        Uint8List? fileNameBytes = msg.readBytes(fileNameLength);
        fields[MessageFields.fileName] = fileNameBytes;
        break;
      case null:
        msg.readBytes(msg.readInt32());
        break;
    }

    int usernameLength = msg.readInt32();
    Uint8List usernameBytes = msg.readBytes(usernameLength);
    String username = utf8.decode(usernameBytes);

    int clientID = msg.readInt32();
    int messageID = msg.readInt32();
    int chatSessionID = msg.readInt32();

    Message message = Message(
      username: username,
      clientID: clientID,
      fields: fields,
      messageID: messageID,
      chatSessionID: chatSessionID,
      chatSessionIndex: UserInfoManager.chatSessionIDSToChatSessions[chatSessionID]!.chatSessionIndex,
      epochSecond: epochSecond,
      contentType: contentType,
      deliveryStatus: clientID == UserInfoManager.clientID
          ? MessageDeliveryStatus.serverReceived
          : MessageDeliveryStatus.delivered,
    );

    ChatSession chatSession = UserInfoManager.chatSessionIDSToChatSessions[chatSessionID]!;
    chatSession.messages.add(message);

    _eventBus.fire(MessageReceivedEvent(message, chatSession));
  }
}

File: ErmisClient/Desktop/src/github/koukobin/ermis/desktop_client/main/java/service/client/handlers/ClientMessageHandler.java
/* Copyright (C) 2023 Ilias Koukovinis <ilias.koukovinis@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package github.koukobin.ermis.desktop_client.main.java.service.client.handlers;

import github.koukobin.ermis.common.message_types.ClientContentType;
import github.koukobin.ermis.common.message_types.MessageDeliveryStatus;
import github.koukobin.ermis.desktop_client.main.java.service.client.Events;
import github.koukobin.ermis.desktop_client.main.java.service.client.GlobalMessageDispatcher;
import github.koukobin.ermis.desktop_client.main.java.service.client.UserInfoManager;
import github.koukobin.ermis.desktop_client.main.java.service.client.models.ChatSession;
import github.koukobin.ermis.desktop_client.main.java.service.client.models.Message;
import io.netty.buffer.ByteBuf;

/**
 * @author Ilias Koukovinis
 *
 */
public class ClientMessageHandler implements MessageHandler {

	@Override
	public void handleMessage(ByteBuf msg) {
		Message message = new Message();

		ClientContentType contentType = ClientContentType.fromId(msg.readInt());
		long epochSecond = msg.readLong();

		byte[] text = null;
		byte[] fileNameBytes = null;

		switch (contentType) {
		case TEXT -> {
			text = new byte[msg.readInt()];
			msg.readBytes(text);
		}
		case FILE, IMAGE, VOICE -> {
			fileNameBytes = new byte[msg.readInt()];
			msg.readBytes(fileNameBytes);
		}
		default -> {
			text = "Error: content type not recognized".getBytes();

			// Skip message content
			msg.skipBytes(msg.readInt());
		}
		}

		byte[] usernameBytes = new byte[msg.readInt()];
		msg.readBytes(usernameBytes);
		String username = new String(usernameBytes);

		int clientID = msg.readInt();
		int messageID = msg.readInt();
		int chatSessionID = msg.readInt();

		message.setContentType(contentType);
		message.setUsername(username);
		message.setClientID(clientID);
		message.setMessageID(messageID);
		message.setChatSessionID(chatSessionID);
		message.setChatSessionIndex(
				UserInfoManager.chatSessionIDSToChatSessions.get(chatSessionID).getChatSessionIndex()
		);
		message.setText(text);
		message.setFileName(fileNameBytes);
		message.setEpochSecond(epochSecond);
		message.setDeliveryStatus(MessageDeliveryStatus.DELIVERED);

		ChatSession chatSession = UserInfoManager.chatSessionIDSToChatSessions.get(chatSessionID);

		if (chatSession.haveChatMessagesBeenCached()) {
			chatSession.getMessages().add(message);
		}

		GlobalMessageDispatcher.getDispatcher().messageSubject.onNext(new Events.MessageReceivedEvent(message, chatSession));
	}

}
